pkg/reconciler/clusterbuildtemplate/resources/imagecache_test.go

```diff
@@ -0,0 +1,231 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resources
+
+import (
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/knative/build/pkg/apis/build/v1alpha1"
+	"github.com/knative/build/pkg/system"
+	caching "github.com/knative/caching/pkg/apis/caching/v1alpha1"
+)
+
+func TestMakeImageCache(t *testing.T) {
```

Comment:
Duplicating this test also feels Wrong, if the code was in a single common location then you'd only need to have one test for it.

Reply:
This checks that we properly create things in the system namespace.  I can reduce the testing, but the vast majority of the logic is shared :)

Reply:
Yeah I think what I'm asking for is a test that does all of these cases, for either BuildTemplate or ClusterBuildTemplate, and checks that the resources are created either in the template's namespace or in the case of ClusterBuildTemplate, the system namespace. Sort of like a two-level nested table-driven test, does that make sense?

Reply:
Yes, let me spend some time on this.  Will ping this when RFAL.

Reply:
ok this should be RFAL.